logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamReaderImpl.java

```diff
@@ -0,0 +1,278 @@
+
+  private final LoggedEventImpl nextEvent = new LoggedEventImpl();
+  private final DirectBuffer nextEventBuffer = new UnsafeBuffer(0, 0);
+
+  // state
+  private IteratorState state;
+  private long nextLogStorageReadAddress;
+  private long lastReadAddress;
+  private int bufferOffset;
+
+  public LogStreamReaderImpl(final LogStorage logStorage) {
+    this.storageReader = logStorage.newReader();
+    invalidateBufferAndOffsets();
+    seek(FIRST_POSITION);
+  }
+
+  @Override
+  public boolean seekToNextEvent(final long position) {
+
+    if (position <= -1) {
+      seekToFirstEvent();
+      return true;
+    }
+
+    final boolean found = seek(position);
+    if (found && hasNext()) {
+      next();
+      return true;
+    }
+
+    return false;
+  }
+
+  @Override
+  public boolean seek(final long position) {
+    if (state == IteratorState.CLOSED) {
+      throw new IllegalStateException(ERROR_CLOSED);
+    }
+
+    final long seekAddress = storageReader.lookUpApproximateAddress(position);
+    invalidateBufferAndOffsets();
+    return seekFrom(seekAddress, position);
+  }
+
+  @Override
+  public void seekToFirstEvent() {
+    seek(FIRST_POSITION);
+  }
+
+  @Override
+  public long seekToEnd() {
+    // invalidate events first as the buffer content may change
+    invalidateBufferAndOffsets();
+
+    final long blockAddress = storageReader.getFirstBlockAddress();
+    if (blockAddress < 0) {
+      // no block found => empty log
+      state = IteratorState.EMPTY_LOG_STREAM;
+      return -1;
+    } else {
+      if (readLastBlockIntoBuffer()) {
+        do {
+          nextEvent.wrap(nextEventBuffer, bufferOffset);
+          bufferOffset += nextEvent.getLength();
+        } while (bufferOffset < nextEventBuffer.capacity());
+
+        return nextEvent.getPosition();
+      }
+
+      // if the log is not empty this should not happen however
+      Loggers.LOGSTREAMS_LOGGER.warn("Unexpected non-empty log failed to read the last block");
+      return UNINITIALIZED;
+    }
+  }
+
+  @Override
+  public long getPosition() {
+    // if an event was already returned use it's position otherwise use position of next event if
+    // available, kind of strange but seemed to be the old API
+    if (isReturnedEventInitialized()) {
+      return returnedEvent.getPosition();
+    }
+
+    if (state == IteratorState.EVENT_AVAILABLE) {
+      return nextEvent.getPosition();
+    }
+
+    return UNINITIALIZED;
+  }
+
+  @Override
+  public long lastReadAddress() {
+    return lastReadAddress;
+  }
+
+  private boolean seekFrom(final long blockAddress, final long position) {
+    if (blockAddress < 0) {
+      // no block found => empty log
+      state = IteratorState.EMPTY_LOG_STREAM;
+      return false;
+    } else {
+      readBlockIntoBuffer(blockAddress);
+      readNextEvent();
+      return searchPositionInBuffer(position);
+    }
+  }
+
+  @Override
+  public void close() {
+    nextEventBuffer.wrap(0, 0);
```

Comment:
Do we need to do this for the other buffer as well?

Reply:
Hm, on one hand if we don't it means we keep a reference to the allocated buffer (but I hope the reader is also garbage collected if it is closed, so not too bad). If we do, however, then if someone is still using the `returnedEvent` then any usage will explode. So as it is now it means you can still use the `returnedEvent` after you close the reader - the `returnedEvent` is only made unusable when you do a `next()` call.

I think this is the behaviour I would expect? But maybe it's safer to reset it...not sure. What do you think?